### PR TITLE
Maxtext Offline serverless inference code

### PR DESCRIPTION
**Pull request by vipannalla**
We coded a offline inference solution for mlperf round to make submissions. The README.md included has clear steps on how to run the offline inference code.

### CODE REVIEW

File: MaxText/inference_mlperf/README.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+```
+export LOAD_PARAMS_PATH=gs://${USER}-bkt/llama2-70b-chat/param-only-decode-ckpt-maxtext/checkpoints/0/items
+
+export SAVE_QUANT_PARAMS_PATH=gs://${USER}-bkt/quantized/llama2-70b-chat
+```
+
+2. Run the following maxtext script to generate and save an in8 quantized checkpoint
+
+```
+export TOKENIZER_PATH=maxtext/assets/tokenizer.llama2
```
[the remark this answers is not in the document]
Comment (vipannalla) [reply MaxText/inference_mlperf/README.md:76]: This is just an example for LLama 70B, which is the only model supported on offline inference using loadgen. When we add support to mixtral 8x7B, this doc needs to be updated. I'll add a  clarifying comment.

[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/inference_mlperf/README.md:76]: I was sugesting maybe we can replace these as well: # HF model id
TOKENIZER_PATH="meta-llama/Llama-2-70b-chat-hf"

[the remark this answers is not in the document]
Comment (vipannalla) [reply MaxText/inference_mlperf/README.md:78]: sg

File: MaxText/maxengine.py [hunk 2]
```diff
@@ -445,3 +465,30 @@ def mesh(self) -> jax.sharding.Mesh:
...
+def create_engine_from_config_flags(batch_size, max_prefill_predict_length, max_target_length, args_str):
+  """ Create new MaxEngine instance with given batch_size, prefill and target lengths, and any config
+      params provided through `args_str`.
+  """
+  import copy
+  import pyconfig
+  args_str = "MaxText/maxengine_server.py configs/base.yml " + args_str
+  args = [b for b in args_str.split(' ') if len(b) > 0]
+
+  args.append("scan_layers=false")
```
[the remark this answers is not in the document]
Comment (vipannalla) [reply MaxText/maxengine.py:485]: Adding configs passed in the end so they override the defaults..

**Comment by vipannalla** [comment]
For some reason the linter complains about "R0917: Too many positional arguments" in files/modules which were not modified by the PR. Was this a new rule recently enabled?

File: MaxText/maxengine.py [hunk 3]
```diff
@@ -441,3 +463,45 @@ def mesh(self) -> jax.sharding.Mesh:
...
+
+  # batch and cache related
+  args.append(f"max_prefill_predict_length={max_prefill_predict_length}")
+  args.append(f"max_target_length={max_target_length}")
+  args.append(f"per_device_batch_size={batch_size}")
+
+  args_str = "MaxText/maxengine_server.py configs/base.yml " + args_str
+  cmd_args = [b for b in args_str.split(' ') if len(b) > 0]
+  # Add at the end to cmd args override the default values set above
+  args.extend(cmd_args)
```
Comment (singh-mitali) [inline MaxText/maxengine.py:502]: Can we please test this? Not sure how pyconfig handles duplicate args in list. My preference would be to override them.

File: MaxText/inference_mlperf/llama_offline_accuracy_run.sh [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+if [ -z "$BASEDIR"];
+then
+  BASEDIR=/home/$USER/inference_mlperf4.1
+fi
+
+USER_CONFIG=$BASEDIR/language/llama2-70b/tpu/user.conf
+
+if [ -z "$DATA_DISK_DIR"];
+then
+  DATA_DISK_DIR=/home/$USE$USERR/loadgen_run_data
```
Comment (singh-mitali) [inline MaxText/inference_mlperf/llama_offline_accuracy_run.sh:13]: typo in user.

File: MaxText/maxengine.py (hunk 3 above)
Comment (vipannalla) [reply MaxText/maxengine.py:502]: I already tested the pyconfig duplicate args usecase before updating this code.

**Review by singh-mitali — APPROVED** [review]

**Comment by gobbleturk** [comment]
> For some reason the linter complains about "R0917: Too many positional arguments" in files/modules which were not modified by the PR. Was this a new rule recently enabled?

This should be getting fixed soon, you may need to sync to head once its fixed

**Pull request merged**